### ISSUE TITLE
[BUG] `__serialize_dbobjs__` breaks cached objects when used according to documentation

### ISSUE THREAD

**Issue by InspectorCaracal** [evennia/evennia#3275]
#### Describe the bug
The serialize and deserialize dunders modify the class object in place. When writing objects to the database or retrieving them from the database, this doesn't cause any problems.

*However.* When adding an object to an already cached db attribute such as a `_SaverList`, it will serialize the object in-place *in the deserialized cache* as part of the write process.

As a result, all of the db objects will have been converted to pickled bytes within the cache, where they should be the deserialized db object references.

#### To Reproduce
I have a class GetAction which serializes `actor` and `targets` in `__serialize_dbobjs__` and deserializes them in `__deserialize_dbobjs__`, as expected.

This is what happens:

```
>>> from base_systems.things.actions import GetAction
>>> act = GetAction(self, here.contents[-1:])
>>> self.db.temporary = list()
>>> print(act.actor)
Monty
>>> self.db.temporary.append(act)
>>> print(act.actor)
b'x80x02(Xx10x00x00x00__packed_dbobj__qx00Xx07x00x00x00objectsqx01Xx08x00x00x00objectdbqx02x86qx03Xx1ax00x00x002023:09:15-14:51:58:955692qx04K(tqx05.'
```

#### Expected behavior
I don't know what the solution is, but supporting serialization of attributes to the db should not break in-memory objects like this.

#### Environment, Evennia version, OS etc
         Evennia version: 2.3.0
         OS: nt
         Python: 3.10.9
         Twisted: 22.10.0
         Django: 4.1.7

**Comment by ChrisLR** [comment]
As proposed in the discord, a good alternative would be for the method to provide a serializable copy instead.
We should probably make a new method for it to avoid breaking compatibility but still change the docs to the new method and `deprecate` the old one.

**Comment by Griatch** [comment]
This is a tricky one. I'd need to investigate further. It's also possible that there's a point where nesting essentially un-pickle-able things ever deeper will just not be supported :/ 

Not sure what @ChrisLR refers to here; note that the return of  `__serialiaze_dbobjs__` is under control of the developer. What meethod should return a serializable copy and what method should be deprecated?

**Comment by ChrisLR** [comment]
@Griatch referring to this 
https://www.evennia.com/docs/latest/Components/Attributes.html#id7
If I understand correctly we are expected to make the changes in-place before it is serialized and then revert them once the serialization is done.

My proposal is the serializer would instead call _as_serializable() or similar on the object,
which would return a serializable copy of the object.
This avoids any in-place changes to the object being used.

**Comment by Griatch** [comment]
@ChrisLR Oh, you mean returning a copy of the _entire_ instance, along with ready serialized variables inside it? Something like this?

```py
class Foo:

    def __init__(self, x, y): 
        # examples of un-pickleable 'hidden' properties
        self.x = Account.objects.get(id=x) 
        self.y = Character.objects.get(id=y) 

    def __as_serializable__(self): 
        copy_foo = Foo(self.x, self.y) 
        copy_foo.x = dbserialize(foo.x)
        copy_foo.y = dbserialize(foo.y) 
        return copy_foo 
```

**Comment by ChrisLR** [comment]
@Griatch yes that could work.
It could also be a thin copy, so you don't have to include everything.
Maybe even a dict at that point....
Advantages it the dev has more control over how serialization is done and what is serialized.
We'd also implement the opposite, a classmethod  `__from_serializable__(cls, value)` that returns the proper value...

.... .... this would also bring us closer to being able to postpone writing to the db but thats another talk. :laughing:

**Comment by Griatch** [comment]
Hm. Simplifying it to a dict is an interesting idea. I think for something like this to work, one would need to invent a new protocol for storing entities of this type. 

1. Serialization: The class instance `instance.__to_serializae__` is called. This returns a 'safe to pickle' version of the class, where the unsafe hidden variables have been replaced with dummy strings, for example. It also returns the 'unsafe' variable names like `{"key": value}`. The serializer will group these together into a structure like the below, and since it's all simple structures now, Evennia will know to properly iterate over it and pickle all components properly: 

    ```python
    {
        "__serialized_class__": <safe-to-pickle class instance> 
       "__serialized_variables__": 
            {
                "variable_name1": <un-pickleable object (like a dbobject) previously hidden on class>,
                "variable_name2:" <un-pickleable object (like a dbobject) previously hidden on class>,
               ...
            }
    }
    ```
2. Deserialization: The above storage structure is depickled, and the deserialized class instance's `__from_serialize__` method is called with the already ready-to-go 'unsafe variables' that the method can now put back in place for the dummies. We should now have a working class instance back.

Not sure if there would be any weird side effects from this apart from de/serialization overhead, but it sounds like it could work.
